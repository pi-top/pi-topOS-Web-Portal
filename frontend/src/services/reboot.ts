import axios from "axios";

import apiBaseUrl from "./apiBaseUrl";

export default async function reboot() {
  await axios.post(
    `${apiBaseUrl}/reboot`,
    {}
  );
}
