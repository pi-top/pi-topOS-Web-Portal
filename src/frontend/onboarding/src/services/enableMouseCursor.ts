import axios from "axios";

import apiBaseUrl from "./apiBaseUrl";

export default async function enableMouseCursor() {
    await axios.post(
        `${apiBaseUrl}/enable-mouse-cursor`,
        {}
    );
}
