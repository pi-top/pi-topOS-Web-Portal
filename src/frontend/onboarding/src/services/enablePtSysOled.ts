import axios from "axios";

import apiBaseUrl from "./apiBaseUrl";

export default async function enablePtSysOled() {
    await axios.post(
        `${apiBaseUrl}/enable-pt-sys-oled`,
        {}
    );
}
