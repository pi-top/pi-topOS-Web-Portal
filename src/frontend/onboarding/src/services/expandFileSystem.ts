import axios from "axios";

import apiBaseUrl from "./apiBaseUrl";

export default async function expandFileSystem() {
  await axios.post(
    `${apiBaseUrl}/expand-fs`,
    {}
  );
}
