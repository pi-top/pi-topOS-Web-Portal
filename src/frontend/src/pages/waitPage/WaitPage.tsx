import React from "react";

import styles from "./WaitPage.module.css";
import Layout from "../../components/layout/Layout";
import Spinner from "../../components/atoms/spinner/Spinner";

export type Props = {
  hasError: boolean,
  isCheckingFsStatus: boolean,
  isExpandingFs: boolean,
  isRebooting: boolean
};

export enum ExplanationMessage {
  CheckingFsStatus = "",
  ExpandingFs = "Expanding file system",
  Rebooting = "Rebooting",
  Default = ""
}

export enum ErrorMessage {
  GenericError = "There was an issue, please turn your device off and back on.",
}

export default ({ isCheckingFsStatus, isExpandingFs, isRebooting, hasError }: Props) => {
  const errorMessage = hasError && ErrorMessage.GenericError;
  const getExplanation = () => {
    if (isRebooting) {
      return ExplanationMessage.Rebooting;
    }
    if (isExpandingFs) {
      return ExplanationMessage.ExpandingFs;
    }
    if (isCheckingFsStatus) {
      return ExplanationMessage.CheckingFsStatus;
    }
    return ExplanationMessage.Default;
  }
  return (
    <Layout
      banner={{
        src: "",
        alt: "wait-screen"
      }}
      prompt={
        (isExpandingFs || isRebooting) && <>
          <Spinner size={50} />
        </>
      }
      explanation={getExplanation()}
      nextButton={{
        onClick: () => {},
        label: '',
        hidden: true
      }}
      className={styles.root}
    >
      {errorMessage && <span className={styles.error}>{errorMessage}</span>}
    </Layout>
  );
};
