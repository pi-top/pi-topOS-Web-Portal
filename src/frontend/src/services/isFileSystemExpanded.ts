import axios from "axios";

import apiBaseUrl from "./apiBaseUrl";

export default async function isFileSystemExpanded() {
  const { data } = await axios.get<{expanded: boolean}>(
    `${apiBaseUrl}/is-fs-expanded`
  );

  return data;
}
