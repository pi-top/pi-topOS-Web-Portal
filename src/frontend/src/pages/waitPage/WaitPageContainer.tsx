import React, { useState, useEffect } from "react";

import WaitPage from "./WaitPage";

import isFileSystemExpanded from "../../services/isFileSystemExpanded";
import reboot from "../../services/reboot";
import expandFileSystem from "../../services/expandFileSystem";
import enableMouseCursor from "../../services/enableMouseCursor";

export type Props = {
  goToNextPage: () => void;
};

export default ({
  goToNextPage,
}: Props) => {
  const [isFsExpanded, setIsFsExpanded] = useState(false);
  const [isCheckingFsStatus, setIsCheckingFsStatus] = useState(false);
  const [isExpandingFs, setIsExpandingFs] = useState(false);
  const [isRebooting, setIsRebooting] = useState(false);
  const [error, setError] = useState(false);

  const fileSystemExpansionStatus = () => {
    setError(false);
    setIsCheckingFsStatus(true);

    isFileSystemExpanded()
      .then((isFileSystemExpandedResponse) => {
          setIsFsExpanded(isFileSystemExpandedResponse.expanded);
          if (isFileSystemExpandedResponse.expanded === false) {
              expandFsAndReboot();
          }
        })
      .catch(() => {
            setIsFsExpanded(false);
            setError(true);
        })
        .finally(() => setIsCheckingFsStatus(false))
  };

  const expandFsAndReboot = () => {
    setIsExpandingFs(true);
    expandFileSystem()
        .then(() => {
            return enableMouseCursor()
        })
        .then(() => {
            setIsRebooting(true);
            return reboot()
        })
        .catch(() => setError(true))
  };

  useEffect(() => {
    Promise.all([fileSystemExpansionStatus()]);
  }, []);

  useEffect(() => {
    if (isFsExpanded) {
        goToNextPage();
    };
  }, [isFsExpanded, goToNextPage])

  return (
    <WaitPage
        hasError={error}
        isCheckingFsStatus={isCheckingFsStatus}
        isExpandingFs={isExpandingFs}
        isRebooting={isRebooting}
    />
    );
};
