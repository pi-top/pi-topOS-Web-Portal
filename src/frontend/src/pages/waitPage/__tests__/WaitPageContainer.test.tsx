import React from "react";
import {
  render,
  RenderResult,
  waitForElement,
  wait,
  queryByText,
  getByText,
} from "@testing-library/react";


import WaitPageContainer, { Props } from "../WaitPageContainer";
import querySpinner from "../../../../test/helpers/querySpinner";
import { ExplanationMessage, ErrorMessage } from "../WaitPage";

import isFileSystemExpanded from "../../../services/isFileSystemExpanded";
import reboot from "../../../services/reboot";
import expandFileSystem from "../../../services/expandFileSystem";
import enableMouseCursor from "../../../services/enableMouseCursor";
import enablePtSysOled from "../../../services/enablePtSysOled";

jest.mock("../../../services/isFileSystemExpanded");
jest.mock("../../../services/reboot");
jest.mock("../../../services/expandFileSystem");
jest.mock("../../../services/enableMouseCursor");
jest.mock("../../../services/enablePtSysOled");

const isFileSystemExpandedMock = isFileSystemExpanded as jest.Mock;
const rebootMock = reboot as jest.Mock;
const expandFileSystemMock = expandFileSystem as jest.Mock;
const enableMouseCursorMock = enableMouseCursor as jest.Mock;
const enablePtSysOledMock = enableMouseCursor as jest.Mock;

describe("WaitPageContainer", () => {
  let defaultProps: Props;
  let mount: (props?: Props) => RenderResult;

  beforeEach(async () => {
    isFileSystemExpandedMock.mockResolvedValue({ expanded: false });
    rebootMock.mockResolvedValue("OK");
    expandFileSystemMock.mockResolvedValue("OK");
    enableMouseCursorMock.mockResolvedValue("OK");
    enablePtSysOledMock.mockResolvedValue("OK");

    defaultProps = {
      goToNextPage: jest.fn()
    };

    mount = (props = defaultProps) => {
      return render(<WaitPageContainer {...props} />);
    };
  });

  afterEach(() => {
    isFileSystemExpandedMock.mockRestore();
    rebootMock.mockRestore();
    expandFileSystemMock.mockRestore();
    enableMouseCursorMock.mockRestore();
  });

  it("renders the correct banner image", async () => {
    const { queryByAltText } = mount();
    await wait();

    expect(queryByAltText("wait-screen")).toMatchSnapshot();

    await wait();
  });

  it("renders prompt correctly", async () => {
    const { container: upgradePage } = mount();
    await wait();

    const prompt = upgradePage.querySelector(".prompt");
    expect(prompt).toMatchSnapshot();

    await wait();
  });

  it("Next button is hidden", async () => {
    const { getByText } = mount();

    expect(getByText("Next")).toHaveProperty("hidden", true);

    await wait();
  });

  it("Skip button is not present", async () => {
    const { queryByText } = mount();

    expect(queryByText("Skip")).not.toBeInTheDocument();

    await wait();
  });

  it("doesn't render spinner when checking if file system was expanded", async () => {
    const { container: waitPage } = mount();

    expect(querySpinner(waitPage)).not.toBeInTheDocument();

    await wait();
  });

  it("checks if the file system was expanded", async () => {
    const { } = mount();
    await wait();

    expect(isFileSystemExpandedMock).toHaveBeenCalled();

    await wait();
  });

  it("attempts to expand file system if it's not already expanded", async () => {
    const { } = mount();
    await wait();

    expect(expandFileSystemMock).toHaveBeenCalled();

    await wait();
  });

  it("renders the spinner when expanding the file system", async () => {
    const { container: waitPage, getByText } = mount();

    await waitForElement(() => querySpinner(waitPage));
    expect(querySpinner(waitPage)).toBeInTheDocument();

    await wait();
  });

  it("attempts to reboot after expanding file system", async () => {
    const { } = mount();
    await wait();

    expect(rebootMock).toHaveBeenCalled();

    await wait();
  });


  it("attempts to reboot", async () => {
    isFileSystemExpandedMock.mockResolvedValue({ expanded: false });

    const { } = mount();
    await wait();

    expect(rebootMock).toHaveBeenCalled();

    await wait();
  });

  it("when the fs was already expanded, goes automatically to the Splash page", async () => {
    isFileSystemExpandedMock.mockResolvedValue({expanded: true});
    const { } = mount();
    await wait();

    expect(defaultProps.goToNextPage).toHaveBeenCalled();
    expect(rebootMock).not.toHaveBeenCalled();
    expect(expandFileSystemMock).not.toHaveBeenCalled();

    await wait();
  });

  it("when the fs was already expanded, doesn't expand fs again", async () => {
    isFileSystemExpandedMock.mockResolvedValue({expanded: true});
    const { } = mount();
    await wait();

    expect(expandFileSystemMock).not.toHaveBeenCalled();

    await wait();
  });

  it("when the fs was already expanded, doesn't reboot", async () => {
    isFileSystemExpandedMock.mockResolvedValue({expanded: true});
    const { } = mount();
    await wait();

    expect(rebootMock).not.toHaveBeenCalled();

    await wait();
  });

  it("when there's an error while checking if fs was expanded, renders the error message", async () => {
    isFileSystemExpandedMock.mockRejectedValue(
      new Error("oh oh, something happened")
    );

    const { getByText } = mount();
    await wait();

    await waitForElement(() => getByText(ErrorMessage.GenericError));
  });

  it("when there's an error while trying to reboot, renders the error message", async () => {
    isFileSystemExpandedMock.mockResolvedValue({expanded: false});
    rebootMock.mockRejectedValue(
      new Error("oh oh, something happened")
    );
    const { getByText } = mount();
    await wait();

    await waitForElement(() => getByText(ErrorMessage.GenericError));
  });

  it("when there's an error while expanding filesystem, doesn't call enableMouseCursor nor reboot", async () => {
    isFileSystemExpandedMock.mockResolvedValue({expanded: false});
    expandFileSystemMock.mockRejectedValue(
      new Error("oh oh, something happened")
    );

    const { getByText } = mount();
    await wait();

    expect(enableMouseCursorMock).not.toHaveBeenCalled();
    expect(rebootMock).not.toHaveBeenCalled();

  });

  it("when there's an error while enabling mouse cursor, doesn't call reboot", async () => {
    isFileSystemExpandedMock.mockResolvedValue({expanded: false});
    enableMouseCursorMock.mockRejectedValue(
      new Error("oh oh, something happened")
    );

    const { getByText } = mount();
    await wait();

    expect(rebootMock).not.toHaveBeenCalled();

  });
});
