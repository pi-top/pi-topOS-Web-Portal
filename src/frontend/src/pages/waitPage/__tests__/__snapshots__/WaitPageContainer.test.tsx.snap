// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WaitPageContainer renders prompt correctly 1`] = `
<h1
  class="prompt"
>
  <div
    aria-busy="true"
    class=""
  >
    <svg
      aria-label="audio-loading"
      height="50"
      viewBox="0 0 38 38"
      width="50"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <lineargradient
          id="a"
          x1="8.042%"
          x2="65.682%"
          y1="0%"
          y2="23.865%"
        >
          <stop
            offset="0%"
            stop-color="#00ad9f"
            stop-opacity="0"
          />
          <stop
            offset="63.146%"
            stop-color="#00ad9f"
            stop-opacity=".631"
          />
          <stop
            offset="100%"
            stop-color="#00ad9f"
          />
        </lineargradient>
      </defs>
      <g
        fill="none"
        fill-rule="evenodd"
      >
        <g
          transform="translate(1 1)"
        >
          <path
            d="M36 18c0-9.94-8.06-18-18-18"
            id="Oval-2"
            stroke="#00ad9f"
            stroke-width="2"
          >
            <animatetransform
              attributeName="transform"
              dur="0.9s"
              from="0 18 18"
              repeatCount="indefinite"
              to="360 18 18"
              type="rotate"
            />
          </path>
          <circle
            cx="36"
            cy="18"
            fill="#fff"
            r="1"
          >
            <animatetransform
              attributeName="transform"
              dur="0.9s"
              from="0 18 18"
              repeatCount="indefinite"
              to="360 18 18"
              type="rotate"
            />
          </circle>
        </g>
      </g>
    </svg>
  </div>
</h1>
`;

exports[`WaitPageContainer renders the correct banner image 1`] = `
<img
  alt="wait-screen"
  class="image"
  draggable="false"
  src=""
  style="max-height: 100%; max-width: 100%;"
/>
`;
